src/systemcmds/topic_listener/generate_listener.py

```diff
@@ -124,35 +123,42 @@
 """)
 
 for index, (m, t) in enumerate(zip(messages, topics)):
-	print("void listen_%s(unsigned num_msgs, unsigned topic_instance);" % t)
+	print("void listen_%s(unsigned num_msgs, unsigned topic_instance, unsigned topic_interval);" % t)
 
 print ("""
-\nint listener_main(int argc, char *argv[]) {
-	if(argc < 2) {
-		printf("need at least two arguments: topic name. [optional number of messages to print] [optional instance]\\n");
-		return 1;
+
+\nstatic void listener_generated(char * topic_name, int topic_instance, int topic_rate, int num_msgs) {
+
+""")
+
+print("""
+
+	unsigned topic_interval = 0;
+	if (topic_rate != 0) {
+		topic_interval = 1000 / topic_rate;
 	}
+
 """)
-print("\tunsigned num_msgs = (argc > 2) ? atoi(argv[2]) : 1;")
-print("\tunsigned topic_instance = (argc > 3) ? atoi(argv[3]) : 0;\n")
+
 for index, (m, t) in enumerate(zip(messages, topics)):
 	if index == 0:
-		print("\tif (strncmp(argv[1],\"%s\",50) == 0) {" % t)
+		print("\tif (strncmp(topic_name,\"%s\", %d) == 0) {" % (t, len(t)))
 	else:
-		print("\t} else if (strncmp(argv[1],\"%s\",50) == 0) {" % t)
-	print("\t\tlisten_%s(num_msgs, topic_instance);" % t)
+		print("\t} else if (strncmp(topic_name,\"%s\", %d) == 0) {" % (t, len(t)))
```

Comment:
Can you use `strcmp`? We know that both are null-terminated strings and using `strncmp` can lead to unwanted prefix matches.